nifi-nar-bundles/nifi-jwt-bundle/nifi-jwt-processors/src/main/java/org/apache/nifi/processors/jwt/UnsignJWT.java

```diff
@@ -0,0 +1,236 @@
+
+    public static final Relationship FAILURE_REL = new Relationship.Builder().name("failure").description(
+            "A FlowFile is routed to this relationship if the JWT cannot be validated or unsigned for any reason")
+            .build();
+
+    public static final String HEADER_ATTRIBUTE = "jwt.header";
+    public static final String FOOTER_ATTRIBUTE = "jwt.footer";
+    public static final String JWT_PREFIX_ATTRIBUTE = "jwt.";
+
+    public static final String PUBLIC_KEY_PREFIX = "-----BEGIN PUBLIC KEY-----\n";
+    public static final String PUBLIC_KEY_SUFFIX = "-----END PUBLIC KEY-----";
+    public static final String SSH_RSA_PREFIX = "ssh-rsa ";
+
+    private File publicCerts = null;
+    private boolean jwtInBody = true;
+    private String jwtAttribute;
+    private KeyFactory kf;
+
+    private List<PropertyDescriptor> descriptors;
+    private Set<Relationship> relationships;
+
+    @Override
+    protected void init(final ProcessorInitializationContext context) {
+        final List<PropertyDescriptor> descriptors = new ArrayList<PropertyDescriptor>();
+        descriptors.add(PUBLIC_KEYS_PATH);
+        descriptors.add(JWT_ATTRIBUTE_NAME);
+        this.descriptors = Collections.unmodifiableList(descriptors);
+
+        final Set<Relationship> relationships = new HashSet<Relationship>();
+        relationships.add(SUCCESS_REL);
+        relationships.add(FAILURE_REL);
+        this.relationships = Collections.unmodifiableSet(relationships);
+    }
+
+    @Override
+    public Set<Relationship> getRelationships() {
+        return this.relationships;
+    }
+
+    @Override
+    public final List<PropertyDescriptor> getSupportedPropertyDescriptors() {
+        return descriptors;
+    }
+
+    @OnScheduled
+    public void onScheduled(final ProcessContext context) {
+        // read public cert from property
+        publicCerts = new File(context.getProperty(PUBLIC_KEYS_PATH).getValue());
+        try {
+            kf = KeyFactory.getInstance("RSA");
+        } catch (NoSuchAlgorithmException e) {
+            context.yield();
+        }
+
+        // read attribute from property
+        jwtAttribute = context.getProperty(JWT_ATTRIBUTE_NAME).getValue();
+        if (!"".equals(jwtAttribute)) {
+            jwtInBody = false;
+        }
+    }
+
+    @Override
+    public void onTrigger(final ProcessContext context, final ProcessSession session) throws ProcessException {
+        FlowFile flowFile = session.get();
+        if (flowFile == null) {
+            return;
+        }
+
+        String jwt;
+        if (jwtInBody) {
+            // read JWT from flowfile
+            final ByteArrayOutputStream bytes = new ByteArrayOutputStream();
+            session.exportTo(flowFile, bytes);
+            jwt = bytes.toString();
+        } else {
+            jwt = flowFile.getAttribute(jwtAttribute);
+        }
+
+        for (final File cert : publicCerts.listFiles((dir, name) -> name.toLowerCase().endsWith(".pub"))) {
```

Comment:
The acquisition of the public key should be refactored to a separate method and should be run outside of the `onTrigger()` method for performance reasons. The selected key(s) can be cached rather than being loaded and each attempted for every flowfile that enters the processor.